commit 342ef6fda6dca99ca256c9ca8ad2bd3401f44027
Author: Tim Kosse <tim.kosse@filezillaproject.org>
Date:   Fri Jul 17 10:52:07 2020 +0200

    Fix drawing of scaled bitmaps in wxAuiGenericTabArt and wxAuiSimpleTabArt
    
    When drawing the bitmaps, their content scale factor needs to be taken into
    account: Instead of wxBitmap::GetWidth, wxBitmap::GetScaledWidth needs to be
    used, same with the height.
    
    This change is inspired by commit fc75497794fd6287d479eb92de49a037b5a016d5 from
    master.

diff --git a/src/aui/tabart.cpp b/src/aui/tabart.cpp
index f736f0b6e5..5d8dbf4211 100644
--- a/src/aui/tabart.cpp
+++ b/src/aui/tabart.cpp
@@ -222,9 +222,9 @@ void wxAuiGenericTabArt::SetSizingInfo(const wxSize& tab_ctrl_size,
     int tot_width = (int)tab_ctrl_size.x - GetIndentSize() - 4;
 
     if (m_flags & wxAUI_NB_CLOSE_BUTTON)
-        tot_width -= m_activeCloseBmp.GetWidth();
+        tot_width -= m_activeCloseBmp.GetScaledWidth();
     if (m_flags & wxAUI_NB_WINDOWLIST_BUTTON)
-        tot_width -= m_activeWindowListBmp.GetWidth();
+        tot_width -= m_activeWindowListBmp.GetScaledWidth();
 
     if (tab_count > 0)
     {
@@ -503,7 +503,7 @@ void wxAuiGenericTabArt::DrawTab(wxDC& dc,
     int close_button_width = 0;
     if (close_button_state != wxAUI_BUTTON_STATE_HIDDEN)
     {
-        close_button_width = m_activeCloseBmp.GetWidth();
+        close_button_width = m_activeCloseBmp.GetScaledWidth();
     }
 
     int bitmap_offset = 0;
@@ -514,10 +514,10 @@ void wxAuiGenericTabArt::DrawTab(wxDC& dc,
         // draw bitmap
         dc.DrawBitmap(page.bitmap,
                       bitmap_offset,
-                      drawn_tab_yoff + (drawn_tab_height/2) - (page.bitmap.GetHeight()/2),
+                      drawn_tab_yoff + (drawn_tab_height/2) - (page.bitmap.GetScaledHeight()/2),
                       true);
 
-        text_offset = bitmap_offset + page.bitmap.GetWidth();
+        text_offset = bitmap_offset + page.bitmap.GetScaledWidth();
         text_offset += 3; // bitmap padding
 
     }
@@ -546,8 +546,8 @@ void wxAuiGenericTabArt::DrawTab(wxDC& dc,
         wxRect focusRectBitmap;
 
         if (page.bitmap.IsOk())
-            focusRectBitmap = wxRect(bitmap_offset, drawn_tab_yoff + (drawn_tab_height/2) - (page.bitmap.GetHeight()/2),
-                                            page.bitmap.GetWidth(), page.bitmap.GetHeight());
+            focusRectBitmap = wxRect(bitmap_offset, drawn_tab_yoff + (drawn_tab_height/2) - (page.bitmap.GetScaledHeight()/2),
+                                            page.bitmap.GetScaledWidth(), page.bitmap.GetScaledHeight());
 
         if (page.bitmap.IsOk() && draw_text.IsEmpty())
             focusRect = focusRectBitmap;
@@ -577,7 +577,7 @@ void wxAuiGenericTabArt::DrawTab(wxDC& dc,
             offsetY = 1;
 
         wxRect rect(tab_x + tab_width - close_button_width - 1,
-                    offsetY + (tab_height/2) - (bmp.GetHeight()/2),
+                    offsetY + (tab_height/2) - (bmp.GetScaledHeight()/2),
                     close_button_width,
                     tab_height);
 
@@ -635,14 +635,14 @@ wxSize wxAuiGenericTabArt::GetTabSize(wxDC& dc,
 
     // if the close button is showing, add space for it
     if (close_button_state != wxAUI_BUTTON_STATE_HIDDEN)
-        tab_width += m_activeCloseBmp.GetWidth() + 3;
+        tab_width += m_activeCloseBmp.GetScaledWidth() + 3;
 
     // if there's a bitmap, add space for it
     if (bitmap.IsOk())
     {
-        tab_width += bitmap.GetWidth();
+        tab_width += bitmap.GetScaledWidth();
         tab_width += 3; // right side bitmap padding
-        tab_height = wxMax(tab_height, bitmap.GetHeight());
+        tab_height = wxMax(tab_height, bitmap.GetScaledHeight());
     }
 
     // add padding
@@ -708,15 +708,15 @@ void wxAuiGenericTabArt::DrawButton(wxDC& dc,
     if (orientation == wxLEFT)
     {
         rect.SetX(in_rect.x);
-        rect.SetY(((in_rect.y + in_rect.height)/2) - (bmp.GetHeight()/2));
-        rect.SetWidth(bmp.GetWidth());
-        rect.SetHeight(bmp.GetHeight());
+        rect.SetY(((in_rect.y + in_rect.height)/2) - (bmp.GetScaledHeight()/2));
+        rect.SetWidth(bmp.GetScaledWidth());
+        rect.SetHeight(bmp.GetScaledHeight());
     }
     else
     {
-        rect = wxRect(in_rect.x + in_rect.width - bmp.GetWidth(),
-                      ((in_rect.y + in_rect.height)/2) - (bmp.GetHeight()/2),
-                      bmp.GetWidth(), bmp.GetHeight());
+        rect = wxRect(in_rect.x + in_rect.width - bmp.GetScaledWidth(),
+                      ((in_rect.y + in_rect.height)/2) - (bmp.GetScaledHeight()/2),
+                      bmp.GetScaledWidth(), bmp.GetScaledHeight());
     }
 
     IndentPressedBitmap(&rect, button_state);
@@ -905,9 +905,9 @@ void wxAuiSimpleTabArt::SetSizingInfo(const wxSize& tab_ctrl_size,
     int tot_width = (int)tab_ctrl_size.x - GetIndentSize() - 4;
 
     if (m_flags & wxAUI_NB_CLOSE_BUTTON)
-        tot_width -= m_activeCloseBmp.GetWidth();
+        tot_width -= m_activeCloseBmp.GetScaledWidth();
     if (m_flags & wxAUI_NB_WINDOWLIST_BUTTON)
-        tot_width -= m_activeWindowListBmp.GetWidth();
+        tot_width -= m_activeWindowListBmp.GetScaledWidth();
 
     if (tab_count > 0)
     {
@@ -1066,7 +1066,7 @@ void wxAuiSimpleTabArt::DrawTab(wxDC& dc,
     int close_button_width = 0;
     if (close_button_state != wxAUI_BUTTON_STATE_HIDDEN)
     {
-        close_button_width = m_activeCloseBmp.GetWidth();
+        close_button_width = m_activeCloseBmp.GetScaledWidth();
         text_offset = tab_x + (tab_height/2) + ((tab_width-close_button_width)/2) - (textx/2);
     }
     else
@@ -1110,7 +1110,7 @@ void wxAuiSimpleTabArt::DrawTab(wxDC& dc,
             bmp = m_disabledCloseBmp;
 
         wxRect rect(tab_x + tab_width - close_button_width - 1,
-                    tab_y + (tab_height/2) - (bmp.GetHeight()/2) + 1,
+                    tab_y + (tab_height/2) - (bmp.GetScaledHeight()/2) + 1,
                     close_button_width,
                     tab_height - 1);
         DrawButtons(dc, rect, bmp, *wxWHITE, close_button_state);
@@ -1163,7 +1163,7 @@ wxSize wxAuiSimpleTabArt::GetTabSize(wxDC& dc,
     wxCoord tab_width = measured_textx + tab_height + 5;
 
     if (close_button_state != wxAUI_BUTTON_STATE_HIDDEN)
-        tab_width += m_activeCloseBmp.GetWidth();
+        tab_width += m_activeCloseBmp.GetScaledWidth();
 
     if (m_flags & wxAUI_NB_TAB_FIXED_WIDTH)
     {
@@ -1223,15 +1223,15 @@ void wxAuiSimpleTabArt::DrawButton(wxDC& dc,
     if (orientation == wxLEFT)
     {
         rect.SetX(in_rect.x);
-        rect.SetY(((in_rect.y + in_rect.height)/2) - (bmp.GetHeight()/2));
-        rect.SetWidth(bmp.GetWidth());
-        rect.SetHeight(bmp.GetHeight());
+        rect.SetY(((in_rect.y + in_rect.height)/2) - (bmp.GetScaledHeight()/2));
+        rect.SetWidth(bmp.GetScaledWidth());
+        rect.SetHeight(bmp.GetScaledHeight());
     }
     else
     {
-        rect = wxRect(in_rect.x + in_rect.width - bmp.GetWidth(),
-                      ((in_rect.y + in_rect.height)/2) - (bmp.GetHeight()/2),
-                      bmp.GetWidth(), bmp.GetHeight());
+        rect = wxRect(in_rect.x + in_rect.width - bmp.GetScaledWidth(),
+                      ((in_rect.y + in_rect.height)/2) - (bmp.GetScaledHeight()/2),
+                      bmp.GetScaledWidth(), bmp.GetScaledHeight());
     }
 
 
